### ISSUE TITLE
Refactor URI build/rewrite framework

### ISSUE THREAD

**Issue by kngenie** [internetarchive/wayback#53]
We've gone through several iterations trying to come up with good URL rewrite scheme for archival-URL mode. Our conclusion at this point is that we need to maintain the _form_ of original URL before and after rewrite. By _form_ we mean the various absolute/relative-ness of URL. In other words we want to rewrite full URL to full URL (`http://www.example.com` to `http://web.archive.org/20140101121314/http://www.example.com`), protocol relative to protocol relative (`//www.example.com` to `//web.archive.org/20140101121314/http://www.example.com`), relative path to relative path (`styles/mobile.css` to `styles/mobile.css`) etc.

We found this is rather awkward to achieve with existing framework for URI rewriting. `ResultURIConverter.makeReplayURI()` method takes only two String parameters `datespec` and `url`. Thus it doesn't have access to the context in which `url` was found. To work around this situation, there are a lot of clumsy code around it, which results in overly complex framework. Here are some observations:
- ResultURIConverter is used for building replay URL and rewriting URL. While these look similar, I consider they are distinct services that call for different configuration schemes.
- `ReplayParseContext` has ResultURIConverter instances for each of context flags (ex `cs_`) built through `ContextResultURIConverterFactory`, just for including context flags in replay URL. Those instances would be unnecessary if makeReplayURI() took context flags as an argument.
- `ContextResultURIConverterFactory` has two different uses. While its `getContextConverter` method has single argument called `flags`, implying its context flags, it can also receive replay URL prefix (see `AccessPointAdapter.getUriConverter()`). I suppose the `ContextResultURIConverterFactory` implementations taking context flags would have been unnecessary if `ResultURIConverter.makeReplayURI()` took context flags as an argument.
- `ReplayParseContext.contextualizeUrl(String, String)` checks if URL-rewrite is necessary, and then converts URL to full absolute form before passing it to ResultURIConverter. This makes it impossible for ResultURIConverter implementation to preserve `mode of URL` described above. Considering ResultURIConverter's primary role, these steps should be left to ResultURIConverter implementation. 
- there's one issue in rewriting relative URL: if the URL being replayed does not have path part (ex. `http://www.example.com`, relative URL need to be converted to full path. ResultURIConverter needs to know the URL being replayed to achieve this. This is another support case for additional parameters  in `ResultURIConverter.makeReplayURI()`.
- Memento code (ex. `EmbeddedCDXServerIndex.addTimegateHeaders()`) prepends _mementoPrefix_ to URI returned by `ResultURIConverter` to ensure Memento URLs are always in absolute form. This is necessary because ResultURIConverter is used for two different purposes and breaks if ResultURIConverter returns different forms of URL depending on the context.
- There's no easy way of passing `X-Forwarded-Proto` request header field to URL rewriting so that it can build absolute URL with appropriate protocol (`http` or `https`). We worked around it by storing the header value in `ThreadLocal`.

Our JIRA `ARI-4033` depends on the resolution of this issue.

Resolution Plan:
- Make it clear ResultURIConverter is for constructing _replay_ URL from Capture (full URL and timestamp) and context information (only thing known at present is context flag). I know the class name doesn't represent this role well, but that's >90% use of this interface, currently, and renaming will have widespread effect. We could rename the interface later.
- Have `AccessPointAdapter` implement `ResultURIConverter`. This (along with the change below) should make `ContextResultURIConverterFactory` unnecessary.
- Define new interface for customizing URL rewrite, that receives more information than current ResultURIConverter does.
- Define an interface for passing context of URL (Capture being replayed, baseURL) that `ReplayParseContext` can implement (for better modularity and ease of testing)

**Comment by kngenie** [comment]
To address our internal regression issue (ARI-4184), I went ahead and had ResultURIConverter implementations also implement ContextResultURIConverterFactory interface (and some related change in two ReplayRenderer implementations). This made customizing ArchivalUrlResultURIConverter much easier (fb6f419). It also made AccessPoint.converterFactory no longer necessary.

**Comment by kngenie** [comment]
First round of work committed: bbb7c48.
This change introduces two new interface `ReplayURIConverter` and `ReplayURLTransformer`, and deprecates `ContextResultURIConverterFactroy` and its related methods in `ReplayParseContext`. Interface compatibility is maintained - existing Wayback deployments are supposed to work without configuration/code change.

**Comment by kngenie** [comment]
aafab88: Introduced `AccessPoint.decorateURIConverter()` for per-request override of URL-rewrite rule (clearner solution to ARI-3930: changing protocol by X-Forwarded-Proto header field).
9e1467e: Redesigned interface for cleaner customization.
Merged to master in 7145f52.
Created branch `uriconverter` at the tip of `ARI-4033`.

**Comment by Arkiver2** [comment]
Currently all URLs in the sources of the pages are rewritten as relative URLs. This causes issues when these urls are used as arguments for SWF objects. An example:
*Original:*
<pre><code>    var flashvars = {
		file:		'http://video.xfire.com/62427b.mp4',
		plugins:	'http://media.xfire.com/swf/ova.swf',
		config:		'http://media.xfire.com/content/videocfg2.xml',
	};</pre></code>*Rewritten as:*
<pre><code>    var flashvars = {
		file:		'/web/20150704210250/http://video.xfire.com/62427b.mp4',
		plugins:	'/web/20150704210250/http://media.xfire.com/swf/ova.swf',
		config:		'/web/20150704210250/http://media.xfire.com/content/videocfg2.xml',
	};</pre></code>
This causes issues, because the URLs are later used as arguments for a SWF object to playback a video:
<pre><code>    swfobject.embedSWF('/web/20150704210250/http://media.xfire.com/swf/jwplayer.swf','container1','500','430','9.0.115','false',
		flashvars, params, attributes);</pre></code>
The SWF object is not able to use the relative URLs to find the video and other URLs needed for playback, so the video can't be played.
To make this work absolute URLs should not be rewritten as relative URLs, but as absolute URLs. Example:
*Original:*
<pre><code>    var flashvars = {
		file:		'http://video.xfire.com/62427b.mp4',
		plugins:	'http://media.xfire.com/swf/ova.swf',
		config:		'http://media.xfire.com/content/videocfg2.xml',
	};</pre></code>*Should be rewritten as:*
<pre><code>    var flashvars = {
		file:		'https://web.archive.org/web/20150704210250/http://video.xfire.com/62427b.mp4',
		plugins:	'https://web.archive.org/web/20150704210250/http://media.xfire.com/swf/ova.swf',
		config:		'https://web.archive.org/web/20150704210250/http://media.xfire.com/content/videocfg2.xml',
	};</pre></code>
This problem likely causes problem on not only Xfire.com and makes playback of videos on many archived websites not possible.

Source of the code used in this Issue:
*Rewritten:* https://web.archive.org/web/20150704210250/http://crash.xfire.com/video/62427b/
*Original:* https://web.archive.org/web/20150704210250im_/http://crash.xfire.com/video/62427b/

**Comment by kngenie** [comment]
@Arkiver2 Yes! That's exactly one of the goals of this redesign. New [`ArchivalUrlReplayURIConverter`](https://github.com/internetarchive/wayback/blob/master/wayback-core/src/main/java/org/archive/wayback/archivalurl/ArchivalUrlReplayURIConverter.java) tries to maintain URL style across rewrite - absolute to absolute, and host-relative to host-relative.
